api/v1alpha4/cluster_webhook.go

```diff
@@ -86,6 +103,111 @@ func (c *Cluster) validate() error {
 		)
 	}
 
+	// If the Cluster uses a managed topology.
+	if c.Spec.Topology != nil {
+		// class should be defined.
+		if len(c.Spec.Topology.Class) == 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					field.NewPath("spec", "topology", "class"),
+					c.Spec.Topology.Class,
+					"can't be empty",
+				),
+			)
+		}
+
+		// version should be valid.
+		if !version.KubeSemver.MatchString(c.Spec.Topology.Version) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "topology", "version"), c.Spec.Topology.Version, "must be a valid semantic version"))
+		}
+
+		// MachineDeployment names must be unique.
+		if c.Spec.Topology.Worker != nil {
+			names := make(map[string]struct{})
+			for _, md := range c.Spec.Topology.Worker.MachineDeployments {
+				if _, ok := names[md.Name]; ok {
+					allErrs = append(allErrs,
+						field.Invalid(field.NewPath("ClusterClass", "spec", "topology", "worker", "deployments"), md, fmt.Sprintf("MachineDeployment class %q is defined more than once.", md.Name)),
+					)
+				}
+				names[md.Name] = struct{}{}
+			}
+		}
+
+		switch old {
+		case nil: // On create
+			// c.Spec.InfrastructureRef and c.Spec.ControlPlaneRef could not be set
+			if c.Spec.InfrastructureRef != nil {
+				allErrs = append(
+					allErrs,
+					field.Invalid(
+						field.NewPath("spec", "infrastructureRef"),
+						c.Spec.InfrastructureRef,
+						"can't be set when a Topology is defined",
+					),
+				)
+			}
+			if c.Spec.ControlPlaneRef != nil {
+				allErrs = append(
+					allErrs,
+					field.Invalid(
+						field.NewPath("spec", "controlPlaneRef"),
+						c.Spec.ControlPlaneRef,
+						"can't be set when a Topology is defined",
+					),
+				)
+			}
+
+		default: // On update
+			// class could not be mutated.
+			if c.Spec.Topology.Class != old.Spec.Topology.Class {
+				allErrs = append(
+					allErrs,
+					field.Invalid(
+						field.NewPath("spec", "topology", "class"),
+						c.Spec.Topology.Class,
+						"class can't be mutated",
+					),
+				)
+			}
+
+			// Version could only be increased.
```

Comment:
Should we not support any kind of rollback?

Reply:
That was called out in the proposal, but I don't remember details. 
Might be @srm09 can help here

Reply:
I'm not 100% sure but I thought the only way Kubernetes supports downgrades is via etcd backup, create + etcd restore. Or maybe this only applies to minor updates and we're talking about patch downgrades?

Reply:
I'll move discussion about rollback to a separated issue, this PR is already getting bigger than expected, so let's stick to the proposal...

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/4939